clients/src/main/java/org/apache/kafka/common/metrics/Metrics.java

```diff
@@ -136,11 +139,55 @@ public synchronized Sensor sensor(String name, MetricConfig config, Sensor... pa
         if (s == null) {
             s = new Sensor(this, name, parents, config == null ? this.config : config, time);
             this.sensors.put(name, s);
+            if (parents != null) {
+                for (Sensor parent : parents) {
+                    List<Sensor> children = childrenSensors.get(parent.name());
+                    if (children == null) {
+                        children = new ArrayList<>();
+                        childrenSensors.put(parent.name(), children);
+                    }
+                    children.add(s);
+                }
+            }
         }
         return s;
     }
 
     /**
+     * Remove a sensor (if it exists), associated metrics and its children.
+     *
+     * @param name The name of the sensor to be removed
+     */
+    public void removeSensor(String name) {
+        List<Sensor> removedSensors = new ArrayList<>();
+
+        synchronized (this) {
+            removeSensor(name, removedSensors);
+        }
+
+        /*
+         * Because `Sensor.metrics` acquires a lock on `sensor` and `Sensor.add` calls methods on `Metrics` that
+         * acquire a lock on the metrics instance, it is important to do the `sensor.metrics` calls below outside
+         * the `synchronized(this)` above.
+         */
+        for (Sensor sensor : removedSensors)
+            for (KafkaMetric metric : sensor.metrics())
+                removeMetric(metric.metricName());
+    }
+
```

Comment:
Step 2 will fail during `MetricName` registration as it has not yet been removed, yes. But this is not exactly worse than what we have in trunk now (where we basically end up throwing an exception after adding a sensor in a very similar way when a Selector is created a second time).

I think the concurrency strategy needs to be revisited within the context of removals and metrics/sensors collection, but I was hoping that could be done as part of KAFKA-2419. What do you think?